pkg/services/secrets/kvstore/service.go

```diff
@@ -0,0 +1,171 @@
+		return err
+	}
+	return s.store.Set(ctx, orgId, namespace, typ, value)
+}
+
+func (s *service) Del(ctx context.Context, orgId int64, namespace string, typ string) error {
+	// wait for the service to be healthy before attempting to access the store
+	if err := s.BasicService.AwaitRunning(ctx); err != nil {
+		return err
+	}
+	return s.store.Del(ctx, orgId, namespace, typ)
+}
+
+func (s *service) Keys(ctx context.Context, orgId int64, namespace string, typ string) ([]Key, error) {
+	// wait for the service to be healthy before attempting to access the store
+	if err := s.BasicService.AwaitRunning(ctx); err != nil {
+		return nil, err
+	}
+	return s.store.Keys(ctx, orgId, namespace, typ)
+}
+
+func (s *service) Rename(ctx context.Context, orgId int64, namespace string, typ string, newNamespace string) error {
+	// wait for the service to be healthy before attempting to access the store
+	if err := s.BasicService.AwaitRunning(ctx); err != nil {
+		return err
+	}
+	return s.store.Rename(ctx, orgId, namespace, typ, newNamespace)
+}
+
+func (s *service) GetAll(ctx context.Context) ([]Item, error) {
+	// wait for the service to be healthy before attempting to access the store
+	if err := s.BasicService.AwaitRunning(ctx); err != nil {
+		return nil, err
+	}
+	return s.store.GetAll(ctx)
+}
+
+func (s *service) starting(ctx context.Context) error {
+	// check if the plugin is installed and enabled
+	if err := EvaluateRemoteSecretsPlugin(ctx, s.pluginsManager, s.cfg); err != nil {
+		s.log.Debug("secrets manager evaluator returned false", "reason", err.Error())
+		// use the sql store if a plugin is not enabeld or installed
+		s.store = s.getSQLStore()
+		return nil
+	}
+
+	// if the plugin is available, start it and use it
+	pluginStore, err := s.getPluginStore(ctx)
+	if err != nil {
+		return err
+	}
+
+	if pluginStore != nil {
+		s.store = pluginStore
+		return nil
+	}
+
+	// if the plugin is not available, use the sql store
+	s.log.Debug("secrets kvstore is using the default (SQL) implementation for secrets management")
+	s.store = s.getSQLStore()
+
+	return nil
+}
+
+func (s *service) getSQLStore() SecretsKVStore {
+	store := NewSQLSecretsKVStore(s.sqlStore, s.secretsService, s.log)
+	return WithCache(store, 5*time.Second, 5*time.Minute)
+}
+
+func (s *service) getPluginStore(ctx context.Context) (SecretsKVStore, error) {
+	namespacedKVStore := GetNamespacedKVStore(s.kvstore)
+
+	// attempt to start the plugin
+	secretsPlugin, err := StartAndReturnPlugin(s.pluginsManager, ctx)
+	if err != nil || secretsPlugin == nil {
+		s.log.Error("failed to start remote secrets management plugin")
+		if isFatal, readErr := IsPluginStartupErrorFatal(ctx, namespacedKVStore); isFatal || readErr != nil {
+			// plugin error was fatal or there was an error determining if the error was fatal
+			s.log.Error("secrets management plugin is required to start -- exiting app")
+			if readErr != nil {
+				return nil, readErr
+			}
+			return nil, err
+		}
```

Comment:
This whole thing seems very brittle vs the returned error indicating whether it's fatal or not, but it's also a direct copy of the existing code so...

It does appear to be the only reason this service depends on the kvstore service, so if we could tidy it up we could remove that dependency.

Reply:
> This whole thing seems very brittle vs the returned error indicating whether it's fatal or not

@DanCech yeah, agreed. it's pretty hard to follow, but i'm not sure i have enough context about what is going here to improve this without breaking things. any suggestions?

> It does appear to be the only reason this service depends on the kvstore service

the namespaced kvstore is also used at the bottom of this function

Reply:
@mmandrus i see some code comments added in #52552 related to removing pieces [in grafana 10](https://github.com/grafana/grafana/blob/b7ed2b02f309874afe33f147f61bf7090febfc46/pkg/services/secrets/kvstore/remote_plugin.go#L149). do you happen to know if the error handling can be simplified here similar to what @DanCech is suggesting?

Reply:
@toddtreece the G10 part should be disregarded at this point, as this still hasn't launched on the AMG side and there are still a lot of secrets that need to be unified.

Truth be told, I wrote this code when I was very new to Grafana and Golang, so I would suggest we clean it up if it's in scope. I like Dan's idea, have `StartAndReturnPlugin()` assess if an error is fatal itself and return a wrapped static error if so. Let me know if you'd like me to take a look at it.

Reply:
@mmandrus 👍 thanks for the context. i'll give it a shot and reach out if i encounter any issues